### DISCUSSION TITLE
Is there a way to prevent circular occurence in Data?

### DISCUSSION THREAD

**Discussion by fd6130** [spatie/laravel-data#234] [Q&A] [answered]
```php

class FooData extends Data
{
      public function __constructor(
             public BarData $bar
      )
      {
      
      }

    public static function fromModel(Foo $model)
    {
          return new self(
             BarData::from($model->bar)
          );
    }

}
```
```php
class BarData extends Data
{
      public function __constructor(
             public FooData $foo
      )
      {
      
      }

    public static function fromModel(Bar $model)
    {
          return new self(
             FooData::from($model->foo)
          );
    }

}
```

**Update 2022-10-24**

From the example above (or other similar situation) Foo and Bar model are bidirectional.

```php

$foo = Foo::find(1);

FooData::from(foo); // will loop forever
```

It will keep looping until exhausted. How can I prevent it? Thanks.

**Comment by rubenvanassche** [comment, marked as answer]
In this case I would not use another `from` call in the custom creation methods since this will probably cause the loop. Manually creating the data object is the best option here.

If it still keeps looping, then take a look at the validation, and manually write out the rules or add a `WithoutValidation` attribute to the property, since our validation rules resolver is probably creating nested validation rules loops.

**Reply by fd6130** [reply]
I'm using the data for json response, not request. So i think validation doesn't help in my case. Perhaps lazy property is the way to prevent the loop?

**Reply by rubenvanassche** [reply]
Hi @fd6130,

I've added a circular dependency test: https://github.com/spatie/laravel-data/blob/f5171d49975733ff593aba41e4e63b27c608b62b/tests/DataTest.php#L2505-L2536

And it is working at the moment, could you provide me a test which is broken so we can maybe fix the problem?

**Reply by fd6130** [reply]
> In this case I would not use another from call in the custom creation methods since this will probably cause the loop. Manually creating the data object is the best option here.

I think you got a point there. But I'm not sure what "Manually creating the data object" mean. It is mean I should create the data object with array for that case?

Currently I can prevent it happens with lazy property. I will only include the property I wanted to display. I think this is the correct way right?

**Reply by rubenvanassche** [reply]
Well if that works than that's probably the best way to go! With manual creation I meant to to create the database object using `new CircData(....)` not suitable for all cases but sometimes it is not possible to use all the features of the data package in these kinds of edge cases.

**Reply by fd6130** [reply]
Thanks. I will mark this as answer.
